### ISSUE TITLE
TSX not correctly mapping imported names.

### ISSUE THREAD

**Issue by pieter-lazzaro** [Microsoft/TypeScript#3928]
When using ES6 style imports the name of exported members aren't correctly mapped when passed as to a JSX element's parameters.

test.tsx
```
/// <reference path="../typings/react-router/react-router.d.ts" />

import * as React from "react";
import {RouteHandler} from "react-router";

export default class Test extends React.Component<any, void> {
  
  constructor(props:any) {
    super(props);
  }
  
  render() {
    return (
      <div>
        <h1>Test</h1>
      	<RouteHandler />
      </div>
    );
  }
}
```

testApp.tsx
```
/// <reference path="../typings/react-router/react-router.d.ts" />
import * as React from 'react';

import {run, RouteHandler, Route, HashLocation} from 'react-router';

import Test from './test';

class App extends React.Component<any, any> {
  
  constructor(props:any) {
    super();
  }
  
  render() {
    return (
      <div>
        <h1>App</h1>
        <RouteHandler />
      </div>
    );
  }
}

var routes = (
  <Route path="/" name="home" handler={App}>
    <Route name="test" handler={Test} />
  </Route>
);

run(routes, HashLocation, (Root) => {
  React.render(<Root/>, document.body);
});
```

`tsc --jsx react --module commonjs test.tsx`

testApp.js
```
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
/// <reference path="../typings/react-router/react-router.d.ts" />
var React = require('react');
var react_router_1 = require('react-router');
var test_1 = require('./test');
var App = (function (_super) {
    __extends(App, _super);
    function App(props) {
        _super.call(this);
    }
    App.prototype.render = function () {
        return (React.createElement("div", null, React.createElement("h1", null, "App"), React.createElement(react_router_1.RouteHandler, null)));
    };
    return App;
})(React.Component);
var routes = (React.createElement(react_router_1.Route, {"path": "/", "name": "home", "handler": App}, React.createElement(react_router_1.Route, {"name": "test", "handler": Test})));
react_router_1.run(routes, react_router_1.HashLocation, function (Root) {
    React.render(React.createElement(Root, null), document.body);
});
```

Test is not being rewritten to test_1.default.

**Comment by s-panferov** [comment]
+1 this is very annoying

**Comment by RyanCavanaugh** [comment]
Should have a fix today

**Comment by ngbrown** [comment]
@pieter-lazzaro have you found this as fixed?  I'm getting errors using react-router in TypeScript, so I tried your example exactly to see if I was doing something, and got the same errors as my code:

```
test.tsx(14,9): error TS2605: JSX element type 'Component<RouteHandlerProp, any>' is not a constructor function for JSX elements.
  Property 'render' is missing in type 'Component<RouteHandlerProp, any>'.
testApp.tsx(17,9): error TS2605: JSX element type 'Component<RouteHandlerProp, any>' is not a constructor function for JSX elements.
testApp.tsx(24,3): error TS2605: JSX element type 'Component<RouteProp, any>' is not a constructor function for JSX elements.
testApp.tsx(25,5): error TS2605: JSX element type 'Component<RouteProp, any>' is not a constructor function for JSX elements.
testApp.tsx(30,16): error TS2605: JSX element type 'Component<RouteProp, any>' is not a constructor function for JSX elements.
```

@RyanCavanaugh  The output is now corrected to `test_1["default"]`, but I'm getting the compiler errors.

I'm using ntypescript built from e1c9d28cb0706f81c14ca95b92fa3e2a223cc60b and the most recent react.d.ts and react-loader.d.ts.

**Comment by mhegazy** [comment]
@ngbrown use the nightly  builds instead of ntypescript.

> npm install -g typescript@next

**Comment by RyanCavanaugh** [comment]
@ngbrown Can you post an example?

Just looking at the error messages, it seems likely you wrote something like this:
```ts
declare var Foo: Component<RouteProp, any>;
let x = <Foo />;
```
This code would be incorrect. The names of JSX elements are class *constructor functions*, not class *instances* -- React is going to `new` them. It's as if you had written:
```ts
class Foo { ... }
let Bar = new Foo();
let x = <Bar />; // <- is going to fail at runtime
```

**Comment by ngbrown** [comment]
@RyanCavanaugh I copy and pasted the `test.tsx` and `testApp.tsx` from the original post.

@mhegazy about `typescript@next`--is it supposed to support `"jsx": "react"` in tsconfig.json?  I'm trying to use the webpack ts-loader and it seems to rely on it being set at the tsconfig.json level instead of passing `--jsx react` on the command line.

Either way, using `typescript@next` (on the command line) instead of `ntypescript` still results in the same error message when running `tsc`.

It's possible the [DefinitelyTyped definitions for react-router](https://github.com/borisyankov/DefinitelyTyped/blob/master/react-router/react-router.d.ts) are incomplete, can you take a look at them and suggest what is missing?

**Comment by RyanCavanaugh** [comment]
Reopening for investigation

**Comment by jbrantly** [comment]
@ngbrown To answer your question about `"jsx": "react"` in tsconfig.json, yes, TypeScript supports that.

**Comment by tomitrescak** [comment]
Is this included in typescript 1.7.5? Am getting this error with ts-loader for webpack, while atom typescript performs everything just fine.
